### PR TITLE
Fix the issue for class_weight used with tf.dataset

### LINKED ISSUE #22275: Class_weight with tf.dataset as input to model.fit will throw an error
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: No Mobile device
- **TensorFlow installed from (source or binary)**:binary
- **TensorFlow version (use command below)**:1.10
- **Python version**:2.7
- **Bazel version (if compiling from source)**: ---
- **GCC/Compiler version (if compiling from source)**:---
- **CUDA/cuDNN version**: cuda-8.0
- **GPU model and memory**: (Titan X and GeForce GTX 1080 )
- **Exact command to reproduce**:

### Describe the problem
I am using` tf.keras `in order to be able to feed the train_data using` tf.dataset` API through model.fit directly. It works fine whenever you didn't pass `class_weight`, but if you pass dict of class_weights, it will throw the following error :

`AxisError: axis 1 is out of bounds for array of dimension 1
`

When I debug the error , I found the error happened exactly at line 531 : 
```
> /usr/local/lib/python2.7/dist-packages/tensorflow/python/keras/engine/training_utils.py(530)standardize_weights()
    528       raise ValueError('`class_weight` not supported for '
    529                        '3+ dimensional targets.')
    530     if y.shape[1] > 1:
-->  531       y_classes = np.argmax(y, axis=1)
    532     elif y.shape[1] == 1:

```
The dimension of y is `TensorShape([Dimension(None), Dimension(8)])
`
So` y.shape[1] > 1` , but the problem y is a **tensor** now not an **numpy array,** that is why it throws the previous error.

So is there any solution to this situation? 

### **EDIT**: Adding sample code to reproduce the error.

```
import os, sys, logging
import numpy as np
import tensorflow as tf
import itertools as itt
logging.basicConfig(level=logging.INFO)


def _int64_feature(value):
  return tf.train.Feature(int64_list= tf.train.Int64List(value= [value]))

def _bytes_feature(value):
  return tf.train.Feature(bytes_list= tf.train.BytesList(value= [value]))

def _float32_feature(value):
  return tf.train.Feature(float_list= tf.train.FloatList(value= [value]))

def tf_records_creating(tfrecord_file):
    logging.info('Creating random tfrecord files for 100 sample')

    labels = np.random.uniform(0, num_classes, total_train).astype(np.int32)
    data = np.random.uniform(0, 255, total_train*224*224*3).reshape(total_train, 224, 224, 3).astype(np.int32)

    writer = tf.python_io.TFRecordWriter(tfrecord_file)

    for idx, (image, label) in enumerate(itt.izip(data, labels)):
        image = image.tostring()
        example = tf.train.Example(features=tf.train.Features(feature={
            'label': _int64_feature(int(label)),
            'image': _bytes_feature(image),
        }))
        writer.write(example.SerializeToString())
    writer.close()
    return

def decode(serialized_example):
  features = tf.parse_single_example(
      serialized_example,
      features={
          'image': tf.FixedLenFeature([], tf.string),
          'label': tf.FixedLenFeature([], tf.int64),
      })

  image = tf.decode_raw(features['image'], tf.float32)
  image.set_shape([224*224*3])

  image=tf.reshape(image, (224,224,3))

  label = tf.cast(features['label'], tf.int32)
  label_categorical = tf.one_hot(label,depth= num_classes, on_value=1,off_value=0,dtype=tf.int32,)
  label_categorical = tf.reshape(label_categorical, [num_classes])
  label_categorical.set_shape([num_classes])

  return image, label_categorical

def data_preparing(tfrecord_file):

    logging.info('Preparing the Training tf.dataset ')
    training_files = [tfrecord_file]
    dataset_train = tf.data.TFRecordDataset(training_files, num_parallel_reads=1)
    dataset_train = dataset_train.apply(tf.contrib.data.shuffle_and_repeat(buffer_size=4 * batch_size))
    dataset_train = dataset_train.map(decode, num_parallel_calls=1)  
    dataset_train = dataset_train.batch(batch_size)
    dataset_train = dataset_train.prefetch(tf.contrib.data.AUTOTUNE)
    return dataset_train

def train_model(tfrecord_file):

    base_model = tf.keras.applications.VGG16(include_top=False, weights='imagenet',
                                                 input_shape=(224, 224, 3), pooling='avg')

    for layer in base_model.layers:
        layer.trainable = False

    logging.info('Building Our Classifier')
    x = base_model.output
    x = tf.keras.layers.Flatten()(x)
    x = tf.keras.layers.Dense(512, activation='relu')(x)
    x = tf.keras.layers.BatchNormalization()(x)
    x = tf.keras.layers.Dropout(0.5)(x)
    x = tf.keras.layers.Dense(num_classes, activation='sigmoid')(x)

    model = tf.keras.models.Model(inputs=base_model.input, outputs=x)
    model.summary()
    opt = tf.keras.optimizers.Adam(lr=0.001)
    model.compile(loss='categorical_crossentropy',
                  optimizer=opt, metrics=['accuracy'])
    dataset_train = data_preparing(tfrecord_file=tfrecord_file)

    weighted_array_train= np.array([0.01557266,0.00867447,0.04579864,0.08275284,0.18281397,
       0.30659676, 0.04686068, 0.31092999])
    class_weight_dict = dict(enumerate(weighted_array_train))

    if using_class_weight == True:
        model.fit(x=dataset_train, epochs=epochs, verbose=1,class_weight = class_weight_dict,
              steps_per_epoch=int(np.ceil(total_train / batch_size)))
    else:
        model.fit(x=dataset_train, epochs=epochs, verbose=1,
              steps_per_epoch=int(np.ceil(total_train / batch_size)))
    return

if __name__== '__main__':

    total_train = 100.
    num_classes= 8
    batch_size = 10
    epochs = 100

    #TODO (1) :Set the path to tfrecord file that we will create it.
    tfrecord_file = '~/train.tfrecords'
    tf_records_creating(tfrecord_file)                # Implement this only one time

    using_class_weight= False                         # if you set this to True, you will produce the error
    train_model(tfrecord_file)
```

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #22275 where class_weight used with tf.dataset will result in:
```
AxisError: axis 1 is out of bounds for array of dimension 1
```

This fix fixes the issue by taking into account that y
could be tensor.

This fix fixes #22275.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by harshini-gadige** [comment]
@yongtang  -  Can this issue be moved to "Ready to Pull" status ? I see it requires reviewer approval. Can you please let me know.

**Comment by yongtang** [comment]
Thanks @harshini-gadige for the help. I think a reviewer approval is still needed.

**Comment by yongtang** [comment]
Ping @pavithrasv, any chance to take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet, @pavithrasv, @mrry: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by pavithrasv** [comment]
Looks like the issue has been fixed as part of the following commit https://github.com/tensorflow/tensorflow/commit/c9049de2515cf8643faefa66cc4aea276d390912#diff-4f0d455edc07c640cbb22c8e39a61dd5

Can we just submit the test case for this?

File: tensorflow/python/keras/engine/training_test.py [hunk 1]
```diff
@@ -2035,6 +2035,26 @@ def test_dataset_input_shape_validation(self):
                                    r'expected (.*?) to have shape \(3,\)'):
         model.train_on_batch(dataset)
 
+  @tf_test_util.run_in_graph_and_eager_modes
+  def test_dataset_with_class_weight(self):
+    model = testing_utils.get_small_functional_mlp(1, 4, input_dim=3)
+    optimizer = RMSPropOptimizer(learning_rate=0.001)
+    loss = 'mse'
+    metrics = ['mae', metrics_module.CategoricalAccuracy()]
```
Comment (pavithrasv) [inline tensorflow/python/keras/engine/training_test.py]: May be skip metrics for this test?

**Comment by yongtang** [comment]
Thanks @pavithrasv for the review. I have updated the PR with only the test included. The metrics has been skipped as well. Please take a look.

File: tensorflow/python/keras/engine/training_test.py (hunk 1 above)
Comment (pavithrasv) [reply tensorflow/python/keras/engine/training_test.py]: Resolving this as metrics have been removed in the new commit.

**Review by pavithrasv — APPROVED** [review]

**Comment by was84san** [comment]
Will it be available in tensor flow 1.13? or it is in the current  1.12 version of tensor flow?

**Comment by yongtang** [comment]
@was84san 1.13 branch was out before this PR I believe. So I assume this fix may only be in 2.0 unless it is cherry-picked by 1.13.

**Comment by felippe-mendonca** [comment]
Even with the changes of the [commit](https://github.com/tensorflow/tensorflow/commit/c9049de2515cf8643faefa66cc4aea276d390912#diff-4f0d455edc07c640cbb22c8e39a61dd5) referenced above, I've got the same problem of issue #22275. I'm using tensorflow 1.12 with keras interfaces and feeding the training with a tf.dataset. Looking into the code of commit, I made some changes in order to put my training to run, e.g. change the np.argmax function [here](https://github.com/tensorflow/tensorflow/blob/c9049de2515cf8643faefa66cc4aea276d390912/tensorflow/python/keras/engine/training_utils.py#L751) to a K.argmax. Another change was on [this](https://github.com/tensorflow/tensorflow/blob/c9049de2515cf8643faefa66cc4aea276d390912/tensorflow/python/keras/engine/training_utils.py#L757) comprehension list, I'd got an error saying that it only works with eager mode.
```
TypeError: Tensor objects are only iterable when eager execution is enabled. To iterate over this tensor use tf.map_fn.
```
I changed to a tf.contrib.lookup.HashTable but I'm sure that it isn't the correct solution since it doesn't handle with classes that aren't on the class_weight dict. The point is: it seems that the solution given on commit doesn't cover all the cases. Am I right or I did something wrong to still get these errors?

**Comment by DanielDIQA** [comment]
If you use model.fit with a tf.data dataset iterator and provide the class_weight dictionary, then this error is thrown: 
`TypeError: Tensor objects are only iterable when eager execution is enabled. To iterate over this tensor use tf.map_fn.`

Probable reason: training_utils.py line 722: 
`weights = np.asarray([class_weight[cls] for cls in y_classes if cls in class_weight])`

Since y_classes is a tf.tensor (not a numpy array) you cannot iterate using "for".

**Comment by was84san** [comment]
@ DaneilDIQA did you tried to enable the eager execution  in your code and see if it will work.

**Pull request merged**